#!/usr/bin/env python
import cartesio_planning.pyplan as cpp
from moveit_ros_planning_interface._moveit_roscpp_initializer import roscpp_init
import query_yes_no as query_yn
import cartesian_interface.affine3


roscpp_init('cartesio_planning', [])

planner = cpp.PlannerClient()


joint_names = ["VIRTUALJOINT_1", "VIRTUALJOINT_2", "VIRTUALJOINT_3", "VIRTUALJOINT_4", "VIRTUALJOINT_5", "VIRTUALJOINT_6", "LHipLat", "LHipSag", "LHipYaw", "LKneePitch", "LAnklePitch", "LAnkleRoll", "RHipLat", "RHipSag", "RHipYaw", "RKneePitch", "RAnklePitch", "RAnkleRoll", "WaistLat", "WaistYaw", "LShSag", "LShLat", "LShYaw", "LElbj", "LForearmPlate", "LWrj1", "LWrj2", "RShSag", "RShLat", "RShYaw", "RElbj", "RForearmPlate", "RWrj1", "RWrj2"]

start_position = [-0.014130067578267716, -0.0003950225138868169, -0.1986007829637569, -0.0006509170597944057, 0.8826449737683223,
                  0.000872636860768782, 0.00028897706287522954, -1.8312116317937361, -0.0006430702571908346, 1.3901233356218243,
                  -0.3815580773111069, -0.0007305196988230341, 0.0002886714460364018, -1.8310894135954228, -0.0006433656662272768,
                  1.3902253281031371, -0.3817822880161119, -0.0007305198085548906, -0.0017565787886696135, 6.49293263958643e-05,
                  -0.027415126795615335, 0.35957581411273704, 0.21939513751962691, -1.1557043020776563, 0.010977362157928367,
                  -0.05852717520634846, 5.8580515206005804e-06, -0.027530704068700118, -0.35857980401440936, -0.2192078206212112,
                  -1.1547169324732134, -0.010928758355845523, -0.058163622523019876, 4.3997414411386765e-06]

# with singularity
goal_position = [0.034024568029280396, -0.07504811373351948, -0.13005339070165617, 0.5389633459902076, 1.1316223514550658, -0.5782745461699126, 0.12017636588623921, -1.3675381173889825, 0.2861684800879465, 2.3004434846336106, -0.5554418427259643, -0.06447166956915328, 0.13446850110139266, -1.2784161487093035, 0.3594922487120925, -1.0567824419490034e-06, 0.15151463334427212, -0.023498357083749467, -0.14791747591957194, 0.02249369930571717, 0.055043039095789156, 0.40445979282440764, 0.44855455623617274, -1.2711413691047815, 0.05843132761777578, -0.15068438165289186, 2.065246981657416e-05, -0.2084342669783355, -0.2603362661793475, -0.18636649374679587, -1.2425577397166045, -0.02686941655774241, -0.12371413442781486, 1.883659982707331e-05]

# without singularity
goal_position = [0.05573729400167113, -0.06947883510411024, -0.20751821206670207, 0.12429667516944151, 0.8849470128966822, -0.022411855389329754, -0.11666467659105043, -0.7627741830617885, -0.2734355503290767, 2.3214736573594834, -0.872665808000081, -0.19540428798391687, -0.03468311724094864, -1.7866746567944674, -0.025411763194786535, 1.496657523972716, -0.5655945029726186, -0.10496579393534855, -0.08098361306596884, -0.07331380995906533, 0.17934555560316112, 0.30786277563783015, 0.3270633311559586, -1.2694988284807516, 0.034189528418359566, -0.12636131675787898, 1.204561182183304e-05, 0.04982332460143657, -0.20438964313392347, -0.16974017674830408, -1.2428230667694045, -0.019438513997855827, -0.11188556609797803, 1.090003907467298e-05]
frames_in_contact = ["TCP_L", "TCP_R", "r_foot_upper_right_link", "r_foot_upper_left_link", "r_foot_lower_right_link", "r_foot_lower_left_link"]

joint_map_start = dict(zip(joint_names, start_position))
joint_map_goal = dict(zip(joint_names, goal_position))

planner.setStartState(joint_map_start)
planner.setGoalState(joint_map_goal)
planner.setContactFrames('SET', frames_in_contact)


max_time = 10.0
planner_type = 'RRTConnect'
interp_time = 0.01

base_link = ['l_sole']
distal_link = ['r_sole']

traj_space = "Joint"

flag_ok = False
while (not flag_ok) :
    planner.callPlanner(max_time, planner_type, interp_time, traj_space, distal_link, base_link)  #traj_space, distal_link, base_link
    flag_ok = query_yn.ask('Is this trajectory good for you?')

mapJointTraj = planner.getJointTrajectory()
# mapCartTraj = planner.getCartesianTrajectory()

print mapJointTraj
# print mapCartTraj.keys()




